builtin/credential/cert/path_login.go

```diff
@@ -134,28 +135,47 @@ func (b *backend) verifyCredentials(req *logical.Request) (*ParsedCert, *logical
 	connState := req.Connection.ConnState
 
 	// Load the trusted certificates
-	roots, trusted := b.loadTrustedCerts(req.Storage)
+	roots, trusted, trustedNonCAs := b.loadTrustedCerts(req.Storage)
 
-	// Validate the connection state is trusted
-	trustedChains, err := validateConnState(roots, connState)
-	if err != nil {
-		return nil, nil, err
-	}
+	// If trustedNonCAs is not empty it means that client had registered a non-CA cert
+	// with the backend.
+	if len(trustedNonCAs) != 0 {
+		// Match the trusted chain with the policy
+		return b.matchNonCAPolicy(connState.PeerCertificates[0], trustedNonCAs), nil, nil
+	} else {
+		// Validate the connection state is trusted
+		trustedChains, err := validateConnState(roots, connState)
+		if err != nil {
+			return nil, nil, err
+		}
 
-	// If no trusted chain was found, client is not authenticated
-	if len(trustedChains) == 0 {
-		return nil, logical.ErrorResponse("invalid certificate or no client certificate supplied"), nil
-	}
+		// If no trusted chain was found, client is not authenticated
+		if len(trustedChains) == 0 {
+			return nil, logical.ErrorResponse("invalid certificate or no client certificate supplied"), nil
+		}
+
+		validChain := b.checkForValidChain(req.Storage, trustedChains)
+		if !validChain {
+			return nil, logical.ErrorResponse(
+				"no chain containing non-revoked certificates could be found for this login certificate",
+			), nil
+		}
 
-	validChain := b.checkForValidChain(req.Storage, trustedChains)
-	if !validChain {
-		return nil, logical.ErrorResponse(
-			"no chain containing non-revoked certificates could be found for this login certificate",
-		), nil
+		// Match the trusted chain with the policy
+		return b.matchPolicy(trustedChains, trusted), nil, nil
 	}
+}
 
-	// Match the trusted chain with the policy
-	return b.matchPolicy(trustedChains, trusted), nil, nil
+// matchNonCAPolicy is used to match the client cert with the registered non-CA
+// policies to establish client identity.
+func (b *backend) matchNonCAPolicy(clientCert *x509.Certificate, trustedNonCAs []*ParsedCert) *ParsedCert {
+	for _, trustedNonCA := range trustedNonCAs {
+		tCert := trustedNonCA.Certificates[0]
+		if tCert.SerialNumber.String() == clientCert.SerialNumber.String() && bytes.Equal(tCert.RawIssuer, clientCert.RawIssuer) {
```

Comment:
I think it should be sufficient to compare the `AuthorityKeyId` rather than the full issuer, and it's a smaller byte array so will be faster. This also has the benefit of allowing a CA certificate to renew with the same key ID, but still uniquely identify the client cert, since any conforming CA will never repeat serials numbers.